File: screenshot/ViewControllers/WebViewController.m
//
//  WebViewController.m
//  screenshot
//
//  Created by Corey Werner on 8/14/17.
//  Copyright © 2017 crazeapp. All rights reserved.
//

#import "WebViewController.h"
#import "Loader.h"
#import "Geometry.h"
#import "NetworkingModel.h"
#import "screenshot-Swift.h"

@import Appsee;
@import WebKit;
@import SpriteKit;

@interface WebViewController () <GameSceneDelegate> {
    BOOL _didViewAppear;
    BOOL _isShorteningUrl;
    BOOL _isShowingGame;
    BOOL _isPlayingGame;
}

@property (nonatomic, strong) UIView *loadingCoverView;
@property (nonatomic, strong) Loader *loader;
@property (nonatomic, strong) UIToolbar *toolbar;

@property (nonatomic) BOOL didLoadInitialPage;

@end

@implementation WebViewController
@dynamic view;


#pragma mark - Life Cycle

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        _toolbarEnabled = YES;
        [self setupBarButtonItems];
        
        _loaderLabelText = @"Loading...";
        
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(applicationDidEnterBackground:) name:UIApplicationDidEnterBackgroundNotification object:nil];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(applicationWillEnterForeground:) name:UIApplicationWillEnterForegroundNotification object:nil];
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _webView = ({
        WebView *view = [[WebView alloc] init];
        view.translatesAutoresizingMaskIntoConstraints = NO;
        view.navigationDelegate = self;
        [self.view addSubview:view];
        [view.topAnchor constraintEqualToAnchor:self.view.topAnchor].active = YES;
        [view.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor].active = YES;
        [view.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor].active = YES;
        [view.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor].active = YES;
        
        [view addObserver:self forKeyPath:NSStringFromSelector(@selector(canGoBack)) options:NSKeyValueObservingOptionNew context:NULL];
        [view addObserver:self forKeyPath:NSStringFromSelector(@selector(canGoForward)) options:NSKeyValueObservingOptionNew context:NULL];
        view;
    });
    
    [self setBarButtonItemsToToolbarIfPossible];
    
    if (self.url) {
        [self loadRequestUrl:self.url];
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    if (!self.didLoadInitialPage) {
        [self showLoadingView];
    }
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    _didViewAppear = YES;
    
    if (!self.didLoadInitialPage) {
        [self.loader startAnimation:LoaderAnimationPoseThenSpin];
    }
    
    [Appsee startScreen:@"WebView"];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    if (_isShowingGame) {
        [AnalyticsTrackers.standard track:@"Game Interrupted" properties:@{@"From": @"User Navigating"}];
    }
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    
    _didViewAppear = NO;
    
    [self hideLoadingView];
}

- (void)applicationDidEnterBackground:(NSNotification *)notification {
    if (self.view.window) {
        [self.loader stopAnimation];
        
        if (_isShowingGame) {
            [AnalyticsTrackers.standard track:@"Game Interrupted" properties:@{@"From": @"App Backgrounding"}];
        }
    }
}

- (void)applicationWillEnterForeground:(NSNotification *)notification {
    if (self.view.window) {
        [self.loader startAnimation:LoaderAnimationSpin];
        
        if (_isShowingGame) {
            [AnalyticsTrackers.standard track:@"Game Resumed" properties:@{@"From": @"App Backgrounding"}];
        }
    }
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    BOOL shouldCallSuper = YES;
    
    if (object == self.webView) {
        if ([keyPath isEqualToString:NSStringFromSelector(@selector(canGoBack))] ||
            [keyPath isEqualToString:NSStringFromSelector(@selector(canGoForward))]) {
            shouldCallSuper = NO;
            
            [self syncToolbarNavigationItems];
        }
    }
    
    if (shouldCallSuper) {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    
    if ([self isViewLoaded]) {
        [self.webView removeObserver:self forKeyPath:NSStringFromSelector(@selector(canGoBack))];
        [self.webView removeObserver:self forKeyPath:NSStringFromSelector(@selector(canGoForward))];
    }
}


#pragma mark - Url

- (void)setUrl:(NSURL *)url {
    _url = url;
    
    if (url && [self isViewLoaded]) {
        [self.webView removeAllBackForwardListItems];
        [self loadRequestUrl:url];
    }
}

- (void)loadRequestUrl:(NSURL *)url {
    self.didLoadInitialPage = NO;
    
    [self.webView loadRequest:[NSURLRequest requestWithURL:url]];
}


#pragma mark - Toolbar

- (UIToolbar *)toolbar {
    if (!_toolbar) {
        UIToolbar *toolbar = [[UIToolbar alloc] init];
        toolbar.frame = CGRectMake(0.f, 0.f, 0.f, [toolbar intrinsicContentSize].height);
        toolbar.translatesAutoresizingMaskIntoConstraints = NO;
        toolbar.hidden = !self.toolbarEnabled;
        [self.view addSubview:toolbar];
        [toolbar.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor].active = YES;
        [toolbar.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor].active = YES;
        [toolbar.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor].active = YES;
        _toolbar = toolbar;
        
        self.webView.scrollView.contentInset = ({
            UIEdgeInsets insets = self.webView.scrollView.contentInset;
            insets.bottom = self.toolbar.bounds.size.height;
            insets;
        });
        
        self.webView.scrollView.scrollIndicatorInsets = ({
            UIEdgeInsets insets = self.webView.scrollView.scrollIndicatorInsets;
            insets.bottom = self.toolbar.bounds.size.height;
            insets;
        });
    }
    return _toolbar;
}

- (void)setToolbarEnabled:(BOOL)toolbarEnabled {
    _toolbarEnabled = toolbarEnabled;
    _toolbar.hidden = !toolbarEnabled;
}

- (void)setupBarButtonItems {
    _backItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"Back"] style:UIBarButtonItemStylePlain target:self action:@selector(backAction)];
    
    _forwardItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"Forward"] style:UIBarButtonItemStylePlain target:self action:@selector(forwardAction)];
    
    [self syncToolbarNavigationItems];
    
    _refreshItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"Refresh"] style:UIBarButtonItemStylePlain target:self action:@selector(refreshAction)];
    
    _shareItem = [self createShareItem];
    
    _safariItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"Safari"] style:UIBarButtonItemStylePlain target:self action:@selector(safariAction)];
    
    UIColor *color = [UIColor crazeRed];
    self.backItem.tintColor = color;
    self.forwardItem.tintColor = color;
    self.refreshItem.tintColor = color;
    self.safariItem.tintColor = color;
}

- (void)setBarButtonItemsToToolbarIfPossible {
    if (self.toolbarEnabled) {
        UIBarButtonItem *fixed = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFixedSpace target:nil action:nil];
        fixed.width = [Geometry padding];
        
        UIBarButtonItem *flexible = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:nil action:nil];
        
        self.toolbar.items = @[self.backItem, fixed, self.forwardItem, fixed, self.refreshItem, flexible, self.shareItem, fixed, self.safariItem];
    }
}

- (UIBarButtonItem *)createShareItem {
    UIBarButtonItem *item;
    
    if (_isShorteningUrl) {
        UIActivityIndicatorView *activityView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhite];
        activityView.color = [UIColor crazeRed];
        [activityView startAnimating];
        
        // Resize the width to the share icon's size to prevent sibling views from jumping
        CGRect rect = activityView.frame;
        rect.size.width = 30.f;
        activityView.frame = rect;
        
        item = [[UIBarButtonItem alloc] initWithCustomView:activityView];
        
    } else {
        item = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAction target:self action:@selector(shareAction)];
    }
    
    item.tintColor = [UIColor crazeRed];
    return item;
}

- (void)updateShareItem {
    _shareItem = [self createShareItem];
    [self setBarButtonItemsToToolbarIfPossible];
}

- (void)syncToolbarNavigationItems {
    self.backItem.enabled = [self.webView canGoBack];
    self.forwardItem.enabled = [self.webView canGoForward];
}


#pragma mark - Web View

- (void)webView:(WKWebView *)webView didStartProvisionalNavigation:(WKNavigation *)navigation {
    if (!self.didLoadInitialPage) {
        [self showLoadingView];
        
        if (_didViewAppear) {
            [self.loader startAnimation:LoaderAnimationPoseThenSpin];
        }
    }
}

- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
    self.didLoadInitialPage = YES;
    
    if (!_isPlayingGame) {
        [self hideLoadingView];
    }
}


#pragma mark - Loading

- (void)showLoadingView {
    if (!self.loadingCoverView) {
        _loadingCoverView = ({
            UIView *view = [[UIView alloc] init];
            view.translatesAutoresizingMaskIntoConstraints = NO;
            view.backgroundColor = [UIColor whiteColor];
            [self.view addSubview:view];
            [view.topAnchor constraintEqualToAnchor:self.topLayoutGuide.bottomAnchor].active = YES;
            [view.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor].active = YES;
            [view.bottomAnchor constraintEqualToAnchor:self.bottomLayoutGuide.topAnchor].active = YES;
            [view.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor].active = YES;
            view;
        });
        
        _loader = ({
            Loader *loader = [[Loader alloc] init];
            loader.translatesAutoresizingMaskIntoConstraints = NO;
            [self.loadingCoverView addSubview:loader];
            [loader.centerXAnchor constraintEqualToAnchor:self.loadingCoverView.centerXAnchor].active = YES;
            [NSLayoutConstraint constraintWithItem:loader attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:self.loadingCoverView attribute:NSLayoutAttributeCenterY multiplier:0.8f constant:0.f].active = YES;
            loader;
        });
        
        CGFloat padding = [Geometry padding];
        
        UILabel *loaderLabel = [[UILabel alloc] init];
        loaderLabel.translatesAutoresizingMaskIntoConstraints = NO;
        loaderLabel.text = self.loaderLabelText;
        loaderLabel.textColor = [UIColor gray6];
        loaderLabel.textAlignment = NSTextAlignmentCenter;
        loaderLabel.font = [UIFont preferredFontForTextStyle:UIFontTextStyleSubheadline];
        [self.loadingCoverView addSubview:loaderLabel];
        [loaderLabel.topAnchor constraintEqualToAnchor:self.loader.bottomAnchor constant:padding].active = YES;
        [loaderLabel.leadingAnchor constraintEqualToAnchor:self.loadingCoverView.leadingAnchor constant:padding].active = YES;
        [loaderLabel.trailingAnchor constraintEqualToAnchor:self.loadingCoverView.trailingAnchor constant:-padding].active = YES;
        
        MainButton *button = [MainButton buttonWithType:UIButtonTypeCustom];
        button.translatesAutoresizingMaskIntoConstraints = NO;
        button.backgroundColor = [UIColor crazeGreen];
        [button setTitle:@"Get More Coins" forState:UIControlStateNormal];
        [button addTarget:self action:@selector(showLoadingGame) forControlEvents:UIControlEventTouchUpInside];
        [self.loadingCoverView addSubview:button];
        [button.bottomAnchor constraintEqualToAnchor:self.loadingCoverView.bottomAnchor constant:-[Geometry extendedPadding]].active = YES;
        [button.centerXAnchor constraintEqualToAnchor:self.loadingCoverView.centerXAnchor].active = YES;
    }
}

- (void)hideLoadingView {
    [self.loadingCoverView removeFromSuperview];
    self.loadingCoverView = nil;
    
    [self.loader stopAnimation];
    self.loader = nil;
    
    if (_isShowingGame) {
        _isShowingGame = NO;
        
        [AnalyticsTrackers.standard track:@"Game Interrupted" properties:@{@"From": @"Page Loading"}];
    }
}


#pragma mark - Game

- (void)showLoadingGame {
    _isShowingGame = YES;
    
    SKView *gameView = [[SKView alloc] init];
    gameView.translatesAutoresizingMaskIntoConstraints = NO;
    gameView.ignoresSiblingOrder = YES;
    [self.loadingCoverView addSubview:gameView];
    [gameView.topAnchor constraintEqualToAnchor:self.loadingCoverView.topAnchor].active = YES;
    [gameView.leadingAnchor constraintEqualToAnchor:self.loadingCoverView.leadingAnchor].active = YES;
    [gameView.bottomAnchor constraintEqualToAnchor:self.loadingCoverView.bottomAnchor].active = YES;
    [gameView.trailingAnchor constraintEqualToAnchor:self.loadingCoverView.trailingAnchor].active = YES;
    
    GameScene *scene = (GameScene *)[GameScene unarchiveFromFile:@"GameScene"];
    scene.gameDelegate = self;
    scene.scaleMode = SKSceneScaleModeAspectFill;
    [gameView presentScene:scene];
    
    [self.loader stopAnimation];
}

- (void)gameSceneDidStartGame:(GameScene *)gameScene {
    _isPlayingGame = YES;
}

- (void)gameSceneDidEndGame:(GameScene *)gameScene {
    _isPlayingGame = NO;
    
    if (self.didLoadInitialPage) {
        [self hideLoadingView];
    }
}


#pragma mark - Actions

- (void)backAction {
    [self.webView goBack];
}

- (void)forwardAction {
    [self.webView goForward];
}

- (void)refreshAction {
    self.webView.URL ? [self.webView reload] : [self setUrl:self.url];
    
    [AnalyticsTrackers.standard track:@"Refreshed webpage" properties:@{@"url": self.url.absoluteString}];
}

- (void)shareAction {
    _isShorteningUrl = YES;
    [self updateShareItem];
    
    [NetworkingModel shortenUrl:self.url completion:^(NSURL * _Nullable url) {
        if (url) {
            UIActivityViewController *controller = [[UIActivityViewController alloc] initWithActivityItems:@[url] applicationActivities:nil];
            [self presentViewController:controller animated:YES completion:nil];
        }
        
        _isShorteningUrl = NO;
        [self updateShareItem];
    }];
    
    [AnalyticsTrackers.standard track:@"Shared webpage" properties:@{@"url": self.url.absoluteString}];
}

- (void)safariAction {
    [[UIApplication sharedApplication] openURL:self.url options:@{} completionHandler:nil];
    
    [AnalyticsTrackers.standard track:@"Opened webpage in Safari" properties:@{@"url": self.url.absoluteString}];
}

@end

File: screenshot/View/WebView/WebViewController.swift
//
//  WebViewController.swift
//  screenshot
//
//  Created by Corey Werner on 12/20/17.
//  Copyright (c) 2017 crazeapp. All rights reserved.
//

import UIKit
import WebKit
import Appsee

@objc protocol WebViewControllerDelegate : NSObjectProtocol {
    func webViewController(_ viewController: WebViewController, declinedInvalidURL url: URL)
}

class WebViewController : BaseViewController {
    let webView = WebView()
    fileprivate var loader: Loader?
    var failLabel:UILabel?
    var tryAgainButton:MainButton?
    weak var delegate: WebViewControllerDelegate?
    
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        
        setupBarButtonItems()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        webView.translatesAutoresizingMaskIntoConstraints = false
        webView.navigationDelegate = self
        view.addSubview(webView)
        webView.topAnchor.constraint(equalTo: view.topAnchor).isActive = true
        webView.leadingAnchor.constraint(equalTo: view.leadingAnchor).isActive = true
        webView.bottomAnchor.constraint(equalTo: view.bottomAnchor).isActive = true
        webView.trailingAnchor.constraint(equalTo: view.trailingAnchor).isActive = true
        webView.addObserver(self, forKeyPath: #keyPath(WebView.canGoBack), options: [.new, .old], context: nil)
        webView.addObserver(self, forKeyPath: #keyPath(WebView.canGoForward), options: [.new, .old], context: nil)
        
        syncToolbar()
        setBarButtonItemsToToolbarIfPossible()
        
        
        let loader = Loader()
        loader.translatesAutoresizingMaskIntoConstraints = false
        loader.color = .crazeRed
        loader.activityView.hidesWhenStopped = true
        self.view.addSubview(loader)
        loader.centerXAnchor.constraint(equalTo: self.view.centerXAnchor).isActive = true
        loader.centerYAnchor.constraint(equalTo: self.view.centerYAnchor).isActive = true
        self.loader = loader
        loader.startAnimation()
        
        loadURL(url)
        
    }
    
   
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        Appsee.startScreen("WebView")
    }
    
    
   
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        var shouldCallSuper = true
        
        if let object = object as? WebView, object == webView {
            if keyPath == #keyPath(WebView.canGoBack) || keyPath == #keyPath(WebView.canGoForward) {
                shouldCallSuper = false
                
                syncToolbarNavigationItems()
            }
        }
        
        if shouldCallSuper {
            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
        }
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
        
        if isViewLoaded {
            webView.removeObserver(self, forKeyPath: #keyPath(WebView.canGoBack))
            webView.removeObserver(self, forKeyPath: #keyPath(WebView.canGoForward))
        }
    }
    
    // MARK: URL
    
    private var isShorteningURL = false
    
    private(set) var url: URL?
    
    func loadURL(_ url: URL?) {
        self.url = url
        
        guard let url = url, isViewLoaded else {
            return
        }
        
        webView.load(URLRequest(url: url))
    }
    
    func rebaseURL(_ url: URL?) {
        webView.removeAllBackForwardListItems()
        loadURL(url)
    }
    
    // MARK: Toolbar
    
    var isToolbarEnabled = true {
        didSet {
            syncToolbar()
        }
    }
    
    fileprivate lazy var toolbar: UIToolbar = {
        let toolbar = UIToolbar()
        toolbar.frame = CGRect(x: 0, y: 0, width: 0, height: toolbar.intrinsicContentSize.height)
        toolbar.translatesAutoresizingMaskIntoConstraints = false
        toolbar.isHidden = !self.isToolbarEnabled
        self.view.addSubview(toolbar)
        toolbar.leadingAnchor.constraint(equalTo: self.view.leadingAnchor).isActive = true
        toolbar.bottomAnchor.constraint(equalTo: self.view.layoutMarginsGuide.bottomAnchor).isActive = true
        toolbar.trailingAnchor.constraint(equalTo: self.view.trailingAnchor).isActive = true
        
        var insets = self.webView.scrollView.contentInset
        insets.bottom = toolbar.bounds.size.height
        self.webView.scrollView.contentInset = insets
        
        insets = self.webView.scrollView.scrollIndicatorInsets
        insets.bottom = toolbar.bounds.size.height
        self.webView.scrollView.scrollIndicatorInsets = insets
        
        return toolbar
    }()
    
    fileprivate func syncToolbar() {
        if isViewLoaded {
            // Don't lazy load the toolbar when initially setting to false
            if !isToolbarEnabled && view.subviews.first(where: { $0 is UIToolbar }) == nil {
                return
            }
            
            toolbar.isHidden = !isToolbarEnabled
        }
    }
    
    private(set) var backItem: UIBarButtonItem!
    private(set) var forwardItem: UIBarButtonItem!
    private(set) var refreshItem: UIBarButtonItem!
    private(set) var shareItem: UIBarButtonItem!
    private(set) var safariItem: UIBarButtonItem!
    
    fileprivate func setupBarButtonItems() {
        backItem = UIBarButtonItem(image: UIImage(named: "Back"), style: .plain, target: self, action: #selector(backAction))
        
        forwardItem = UIBarButtonItem(image: UIImage(named: "Forward"), style: .plain, target: self, action: #selector(forwardAction))
        
        syncToolbarNavigationItems()
        
        refreshItem = UIBarButtonItem(image: UIImage(named: "Refresh"), style: .plain, target: self, action: #selector(refreshAction))
        
        shareItem = createShareItem()
        
        safariItem = UIBarButtonItem(image: UIImage(named: "Safari"), style: .plain, target: self, action: #selector(safariAction))
        
        backItem.tintColor = .crazeRed
        forwardItem.tintColor = .crazeRed
        refreshItem.tintColor = .crazeRed
        safariItem.tintColor = .crazeRed
    }
    
    private func createShareItem() -> UIBarButtonItem {
        let item: UIBarButtonItem
        
        if isShorteningURL {
            let activityView = UIActivityIndicatorView(activityIndicatorStyle: .white)
            activityView.color = .crazeRed
            activityView.startAnimating()
            
            // Resize the width to the share icon's size to prevent sibling views from jumping
            var rect = activityView.frame
            rect.size.width = 30
            activityView.frame = rect
            
            item = UIBarButtonItem(customView: activityView)
            
        } else {
            item = UIBarButtonItem(barButtonSystemItem: .action, target: self, action: #selector(shareAction))
        }
        
        item.tintColor = .crazeRed
        return item
    }
    
    private func setBarButtonItemsToToolbarIfPossible() {
        if isToolbarEnabled {
            let fixed = UIBarButtonItem(barButtonSystemItem: .fixedSpace, target: nil, action: nil)
            fixed.width = .padding
            
            let flexible = UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
            
            toolbar.items = [backItem, fixed, forwardItem, fixed, refreshItem, flexible, shareItem, fixed, safariItem]
        }
    }
    
    func updateShareItem() {
        shareItem = createShareItem()
        setBarButtonItemsToToolbarIfPossible()
    }
    
    fileprivate func syncToolbarNavigationItems() {
        backItem.isEnabled = webView.canGoBack
        forwardItem.isEnabled = webView.canGoForward
    }
    
    // MARK: Toolbar Actions
    
    @objc fileprivate func backAction() {
        webView.goBack()
    }
    
    @objc fileprivate func forwardAction() {
        webView.goForward()
    }
    
    @objc func refreshAction() {
        if webView.url == nil {
            let resetURL = url
            url = resetURL
            
        } else {
            webView.reload()
        }
        
        Analytics.trackRefreshedWebpage(url: url?.absoluteString)
    }
    
    @objc fileprivate func shareAction() {
        guard let url = url else {
            return
        }
        
        isShorteningURL = true
        updateShareItem()
        
        NetworkingPromise.sharedInstance.shorten(url: url) { shortenedURL in
            if let shortenedURL = shortenedURL {
                let controller = UIActivityViewController(activityItems: [shortenedURL], applicationActivities: nil)
                self.present(controller, animated: true, completion: nil)
            }
            
            self.isShorteningURL = false
            self.updateShareItem()
        }
    }
    
    @objc fileprivate func safariAction() {
        guard let url = url else {
            return
        }
        
        UIApplication.shared.open(url, options: [:], completionHandler: nil)
    }
}

extension WebViewController : WKNavigationDelegate {
    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, decisionHandler: @escaping (WKNavigationActionPolicy) -> Void) {
        if let url = webView.url {
            if UIApplication.shared.canOpenURL(url) {
                decisionHandler(.allow)
            }
            else {
                decisionHandler(.cancel)
                Analytics.trackWebViewInvalidUrl(url: url.absoluteString)
                self.delegate?.webViewController(self, declinedInvalidURL: url)
            }
        }
        else {
            decisionHandler(.cancel)
        }
    }
    
    func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
       
    }
    
    func webView(_ webView: WKWebView, decidePolicyFor navigationResponse: WKNavigationResponse, decisionHandler: @escaping (WKNavigationResponsePolicy) -> Void) {
        self.loader?.stopAnimation()
        decisionHandler(.allow)
    }
   
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        self.loader?.stopAnimation()
    }
   
    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        self.failedToLoad()
    }
    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        self.failedToLoad()
    }
    func failedToLoad() {
        self.loader?.stopAnimation()
        
        if let label = self.failLabel {
            label.removeFromSuperview()
        }
        if let button = self.tryAgainButton{
            button.removeFromSuperview()
        }
        let failLabel = UILabel()
        failLabel.translatesAutoresizingMaskIntoConstraints = false
        failLabel.text = "products.related_looks.error.connection".localized
        failLabel.textColor = .gray3
        failLabel.textAlignment = .center
        failLabel.font = UIFont.preferredFont(forTextStyle: .subheadline)
        failLabel.numberOfLines = 0
        failLabel.adjustsFontForContentSizeCategory = true
        self.view.addSubview(failLabel)
        failLabel.bottomAnchor.constraint(equalTo: self.view.centerYAnchor, constant:-.padding).isActive = true
        failLabel.centerXAnchor.constraint(equalTo: self.view.centerXAnchor, constant: -.padding).isActive = true
        
        let button = MainButton()
        button.translatesAutoresizingMaskIntoConstraints = false
        button.backgroundColor = .crazeGreen
        button.setTitle("generic.retry".localized, for: .normal)
        button.addTarget(self, action: #selector(tryAgain(_:)), for: .touchUpInside)
        self.view.addSubview(button)
        button.topAnchor.constraint(equalTo: self.view.centerYAnchor, constant:.padding).isActive = true
        button.centerXAnchor.constraint(equalTo: self.view.centerXAnchor).isActive = true
        self.tryAgainButton = button
        self.failLabel = failLabel

    }
    
    @objc func tryAgain(_ sender:Any) {
        self.tryAgainButton?.removeFromSuperview()
        self.tryAgainButton = nil
        
        self.failLabel?.removeFromSuperview()
        self.failLabel = nil
        self.loader?.startAnimation()
        
        loadURL(url)

    }
    
}

